modules/packages/Sort.chpl

```diff
@@ -968,123 +968,196 @@ module MergeSort {
-    if (hi - lo < minlen) {
-      InsertionSort.insertionSort(Data, comparator=comparator);
-      return;
+      // then Data[b] > piv and Data[c] < piv,
+      // so Data[c] < Data[b] and they are an inversion
+      ShallowCopy.shallowSwap(Data[b], Data[c]);
+      b += 1;
+      c -= 1;
     }
 
-    // find pivot using median-of-3 method
-    if (chpl_compare(Data(mid), Data(lo), comparator) < 0) then
-      Data(mid) <=> Data(lo);
-    if (chpl_compare(Data(hi), Data(lo), comparator) < 0) then
-      Data(hi) <=> Data(lo);
-    if (chpl_compare(Data(hi), Data(mid), comparator) < 0) then
-      Data(hi) <=> Data(mid);
+    // Now put piv back in Data[lo]
+    ShallowCopy.shallowCopy(Data[lo], piv); // leaves piv empty
 
-    const pivotVal = Data(mid);
-    Data(mid) = Data(hi-stride);
-    Data(hi-stride) = pivotVal;
-    // end median-of-3 partitioning
+    // now we are in the state:
+    //   |  =  |  <  |  > |  =   |
+    //    lo    a   c b  d     hi
 
-    var loptr = lo,
-        hiptr = hi-stride;
-    while (loptr < hiptr) {
-      do { loptr += stride; } while (chpl_compare(Data(loptr), pivotVal, comparator) < 0);
-      do { hiptr -= stride; } while (chpl_compare(pivotVal, Data(hiptr), comparator) < 0);
-      if (loptr < hiptr) {
-        Data(loptr) <=> Data(hiptr);
-      }
+    // now place the equal regions in the right places
+    var s, l, h: int;
+
+    // Fix the first = region
+    s = min(a-lo, b-a); // the number of of elements to swap
+    l = lo;
+    h = b-s;
+    while s > 0 {
+      ShallowCopy.shallowSwap(Data[l], Data[h]);
+      l += 1;
+      h += 1;
+      s -= 1;
+    }
+
+    // Fix the second = region
+    var n = hi+1;
+    s = min(d-c, hi-d);
+    l = b;
+    h = n-s;
+    while s > 0 {
+      ShallowCopy.shallowSwap(Data[l], Data[h]);
+      l += 1;
+      h += 1;
+      s -= 1;
     }
 
-    Data(hi-stride) = Data(loptr);
-    Data(loptr) = pivotVal;
+    var eqStart = b-a+lo;
+    var eqEnd = hi-(d-c);
 
-    // TODO -- Get this cobegin working and tested
-    //  cobegin {
-      quickSort(Data[..loptr-stride by stride align lo], minlen, comparator);  // could use unbounded ranges here
-      quickSort(Data[loptr+stride.. by stride align lo], minlen, comparator);
-    //  }
+    return (eqStart, eqEnd);
   }
 
-  /* Non-stridable quickSort */
-  proc quickSort(Data: [?Dom] ?eltType, minlen=16, comparator:?rec=defaultComparator)
-    where !Dom.stridable {
+
+  // Puts lo, mid, and hi in order by swapping.
+  proc order3(Data: [?Dom] ?eltType,
+              lo: int, mid: int, hi: int,
+              comparator) {
+    if (chpl_compare(Data[mid], Data[lo], comparator) < 0) then
+      ShallowCopy.shallowSwap(Data[mid], Data[lo]);
+    if (chpl_compare(Data[hi], Data[lo], comparator) < 0) then
+      ShallowCopy.shallowSwap(Data[hi], Data[lo]);
+    if (chpl_compare(Data[hi], Data[mid], comparator) < 0) then
+      ShallowCopy.shallowSwap(Data[hi], Data[mid]);
+  }
```

Comment:
It seems like all the callers want out of this function is for the median element to end up in `Data[mid]`, which would take at most one swap.  Having `Data[lo] < Data[mid]` doesn't seem like it gets you anything, especially since `partition()` is going to start out by swapping them.  Why not have `order3()` leave the median element in `Data[lo]` to start with?

Or even just return the index of the median element, with no swapping.

Reply:
I've done this